File: Heaps/9)LastStoneWeight/Sol.java
// https://leetcode.com/problems/last-stone-weight/description/

import java.util.*;

class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);

        // Add all stones to the max heap
        for (int stone : stones) {
            maxHeap.add(stone);
        }

        // Smash stones until one or no stones are left
        while (maxHeap.size() > 1) {
            int stone1 = maxHeap.poll(); // Heaviest stone
            int stone2 = maxHeap.poll(); // Second heaviest stone

            if (stone1 != stone2) {
                // If the two stones don't have the same weight, add the difference back to the
                // heap
                maxHeap.offer(Math.abs(stone1 - stone2));
            }
        }

        // If there's a stone left, return its weight, otherwise return 0
        return maxHeap.isEmpty() ? 0 : maxHeap.poll();

    }
}

File: Heaps/9)LastStoneWeight/Sol.js
// https://leetcode.com/problems/last-stone-weight/description/

class MaxHeap {
    constructor() {
        this.heap = [];
    }

    push(val) {
        this.heap.push(val);
        this._heapifyUp();
    }

    pop() {
        if (this.heap.length === 0) return null;
        const max = this.heap[0];
        const last = this.heap.pop();
        if (this.heap.length > 0) {
            this.heap[0] = last;
            this._heapifyDown();
        }
        return max;
    }

    _heapifyUp() {
        let index = this.heap.length - 1;
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (this.heap[index] > this.heap[parentIndex]) {
                [this.heap[index], this.heap[parentIndex]] = [
                    this.heap[parentIndex],
                    this.heap[index],
                ];
                index = parentIndex;
            } else {
                break;
            }
        }
    }

    _heapifyDown() {
        let index = 0;
        while (2 * index + 1 < this.heap.length) {
            let largerChildIndex = 2 * index + 1;
            if (
                2 * index + 2 < this.heap.length &&
                this.heap[2 * index + 2] > this.heap[largerChildIndex]
            ) {
                largerChildIndex = 2 * index + 2;
            }
            if (this.heap[index] < this.heap[largerChildIndex]) {
                [this.heap[index], this.heap[largerChildIndex]] = [
                    this.heap[largerChildIndex],
                    this.heap[index],
                ];
                index = largerChildIndex;
            } else {
                break;
            }
        }
    }
}

function lastStoneWeight(stones) {
    const maxHeap = new MaxHeap();

    // Push all stone weights into the max-heap (negate to simulate max-heap)
    stones.forEach((stone) => maxHeap.push(stone));

    // Simulate the process
    while (maxHeap.heap.length > 1) {
        const stone1 = maxHeap.pop();
        const stone2 = maxHeap.pop();

        if (stone1 !== stone2) {
            maxHeap.push(Math.abs(stone1 - stone2));
        }
    }

    // If there is one stone left, return its weight, otherwise return 0
    return maxHeap.heap.length === 0 ? 0 : maxHeap.pop();
}

// Example usage:
const stones = [2, 7, 4, 1, 8, 1];
console.log(lastStoneWeight(stones)); // Output: 1
